File: professor/rust/src/lab2.rs
use chroma::ChromaHttpClient as HttpClient;
use chroma::ChromaCollection;
use chroma::types::*;
use chroma::types::MetadataValue as MV;
use fastembed::TextEmbedding;

use std::collections::HashMap;

pub async fn embedding_and_metadata(client: &HttpClient, model_borrow: Option<&mut TextEmbedding>) -> Result<(), String> {
    let mut model = match model_borrow {
        Some(m) => m,
        None => { return Err("No model provided".to_string()); }
    };
    
    let collection = match client.create_collection("countries", None, None).await {
        Ok(c) => c,
        Err(err) => {
            eprintln!("Error while getting collection: {}", err);
            return Err(format!("Cannot create collection. {}", err));
        }
    };

    let mut countries_vec: Vec<&str> = vec![];
    countries_vec.push("Poland is a country in Central Europe with Warsaw as its capital.");
    countries_vec.push("Germany is a European country known for its strong economy.");
    countries_vec.push("France is famous for Paris, culture and tourism.");
    countries_vec.push("Italy is known for its history, food and architecture.");
    countries_vec.push("Spain is located in Southern Europe and is popular for tourism.");
    countries_vec.push("United States is a large country in North America with a diverse economy.");
    countries_vec.push("Canada is known for its cold climate and high quality of life.");
    countries_vec.push("Japan is an island country in Asia with advanced technology.");
    countries_vec.push("China is one of the largest countries in the world by population.");
    countries_vec.push("Brazil is the largest country in South America with the Amazon rainforest.");
    
    let mut metadata_vec: Vec<Option<HashMap<String, MV>>> = vec![];
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Europe".into())), 
                ("capital".into(), MV::Str("Warsaw".into())),
                ("currency".into(), MV::Str("PLN".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Europe".into())),
                ("capital".into(), MV::Str("Berlin".into())),
                ("currency".into(), MV::Str("EUR".into()))])));
    metadata_vec.push(Some(HashMap::from(
                [("continent".into(), MV::Str("Europe".into())),
                ("capital".into(), MV::Str("Paris".into())),
                ("currency".into(), MV::Str("EUR".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Europe".into())),
                ("capital".into(), MV::Str("Rome".into())),
                ("currency".into(), MV::Str("EUR".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Europe".into())),
                ("capital".into(), MV::Str("Madrit".into())),
                ("currency".into(), MV::Str("EUR".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("North America".into())),
                ("capital".into(), MV::Str("Washington D.C.".into())),
                ("currency".into(), MV::Str("USD".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("North America".into())),
                ("capital".into(), MV::Str("Ottawa".into())),
                ("currency".into(), MV::Str("CAD".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Asia".into())),
                ("capital".into(), MV::Str("Tokio".into())),
                ("currency".into(), MV::Str("JPY".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("Asia".into())),
                ("capital".into(), MV::Str("Beijing".into())),
                ("currency".into(), MV::Str("CNY".into()))])));
    metadata_vec.push(Some(HashMap::from([
                ("continent".into(), MV::Str("South America".into())),
                ("capital".into(), MV::Str("Brasília".into())),
                ("currency".into(), MV::Str("BRL".into()))])));
    let mut ids_vec: Vec<String> = vec![];
    for i in 0..countries_vec.len() {
        ids_vec.push(i.to_string());
    }
    let embeddings = match model.embed(countries_vec.clone(), None) {
        Ok(em) => em,
        Err(err) => {
            return Err(format!("Error while generating embeddings: {}", err));
        }
    };
    let some_vec: Vec<Option<String>> = countries_vec.iter().map(|text| Some(text.to_string())).collect();
    match collection
        .add(ids_vec,
            embeddings,
            Some(some_vec),
            None,
            Some(metadata_vec),
        ).await {
    Ok(_) => println!("Inserted items successfuly."),
    Err(err) => {
            return Err(format!("Error while inserting to collection: {}", err));
        }
    }
    let query = "Country with strong economy in Europe".to_string();
    let query = match model.embed(vec![query], None) {
        Ok(em) => em,
        Err(err) => {
            return Err(format!("Error while generating embeddings: {}", err));
        }
    };
    
    // Standard search
    let results = match collection
        .query(query, Some(1), None, None, None)
        .await {
            Ok(r) => r,
            Err(err) => {
                return Err(format!("Error while getting results from collection: {}", err));
            }
        };
    println!("Standard result: {:#?}", results);
    let query = "Country".to_string();
    let query = match model.embed(vec![query], None) {
        Ok(em) => em,
        Err(err) => {
            return Err(format!("Error while generating embeddings: {}", err));
        }
    }; 
    // Only Europe
    let results = match collection
        .query(
            query,
            Some(5),
            Some(Where::Metadata(MetadataExpression {
                        key: "continent".to_string(),
                        comparison: MetadataComparison::Primitive(
                                PrimitiveOperator::Equal,
                                MV::Str("Europe".to_string()),
                            )
                    })
                ),
            None,
            None)
        .await {
            Ok(r) => r,
            Err(err) => {
                return Err(format!("Error while getting results from collection: {}", err));
            }
        };
    println!("Continent result: {:#?}", results);
    let query = "Capital city".to_string();
    let query = match model.embed(vec![query], None) {
        Ok(em) => em,
        Err(err) => {
            return Err(format!("Error while generating embeddings: {}", err));
        }
    }; 
    // Only Tokyo
    let results = match collection
        .query(
            query,
            Some(5),
            Some(Where::Metadata(MetadataExpression {
                        key: "capital".to_string(),
                        comparison: MetadataComparison::Primitive(
                                PrimitiveOperator::Equal,
                                MV::Str("Tokyo".to_string()),
                            )
                    })
                ),
            None,
            None)
        .await {
            Ok(r) => r,
            Err(err) => {
                return Err(format!("Error while getting results from collection: {}", err));
            }
        };
    println!("Tokyo result: {:#?}", results);
    let query = "Big country with large population in Asia".to_string();
    let query = match model.embed(vec![query], None) {
        Ok(em) => em,
        Err(err) => {
            return Err(format!("Error while generating embeddings: {}", err));
        }
    }; 
    // Big countries in Asia
    let results = match collection
        .query(
            query,
            Some(2),
            None,
            None,
            None)
        .await {
            Ok(r) => r,
            Err(err) => {
                return Err(format!("Error while getting results from collection: {}", err));
            }
        };
    println!("Big countires result: {:#?}", results);
    
    Ok(())
}

File: professor/typescript/lab2.ts
import { ChromaClient } from "chromadb";
import { FlagEmbedding, EmbeddingModel } from "fastembed";

async function main() {
    console.log("Ładowanie modelu...");
    const model = await FlagEmbedding.init({ model: EmbeddingModel.BGESmallENV15 });

    // Punkt 1: Inicjalizacja klienta i tworzenie kolekcji
    const client = new ChromaClient();

    try {
        await client.deleteCollection({ name: "countries" });
    } catch (e) {
    }
    const collection = await client.createCollection({ name: "countries" });

    // Punkt 2: Dokumenty
    const documents = [
        "Poland is a country in Central Europe with Warsaw as its capital.",
        "Germany is a European country known for its strong economy.",
        "France is famous for Paris, culture and tourism.",
        "Italy is known for its history, food and architecture.",
        "Spain is located in Southern Europe and is popular for tourism.",
        "United States is a large country in North America with a diverse economy.",
        "Canada is known for its cold climate and high quality of life.",
        "Japan is an island country in Asia with advanced technology.",
        "China is one of the largest countries in the world by population.",
        "Brazil is the largest country in South America with the Amazon rainforest."
    ];

    // Punkt 3: Metadane
    const metadatas = [
        { continent: "Europe", capital: "Warsaw", currency: "PLN" },
        { continent: "Europe", capital: "Berlin", currency: "EUR" },
        { continent: "Europe", capital: "Paris", currency: "EUR" },
        { continent: "Europe", capital: "Rome", currency: "EUR" },
        { continent: "Europe", capital: "Madrid", currency: "EUR" },
        { continent: "North America", capital: "Washington D.C.", currency: "USD" },
        { continent: "North America", capital: "Ottawa", currency: "CAD" },
        { continent: "Asia", capital: "Tokyo", currency: "JPY" },
        { continent: "Asia", capital: "Beijing", currency: "CNY" },
        { continent: "South America", capital: "Brasília", currency: "BRL" }
    ];

    const ids = documents.map((_, i) => i.toString());


    // Punkt 4: Generowanie embeddingów
    console.log("Generowanie embeddingów dla dokumentów...");
    const embeddingsGenerator = model.embed(documents);
    const embeddings: number[][] = [];
    for await (const batch of embeddingsGenerator) {
        embeddings.push(...batch.map(e => Array.from(e)));
    }

    // Punkt 5: Zapis do bazy ChromaDB
    console.log("Zapis do bazy ChromaDB...");
    await collection.add({
        ids: ids,
        embeddings: embeddings,
        metadatas: metadatas,
        documents: documents
    });

    async function getQueryEmbedding(text: string): Promise<number[]> {
        return Array.from(await model.queryEmbed(text));
    }


    // Punkt 6: Wyszukiwanie semantyczne państwa o podobnym opisie
    let query = "country with strong economy in Europe";
    let queryEmbedding = await getQueryEmbedding(query);

    let results = await collection.query({
        queryEmbeddings: [queryEmbedding],
        nResults: 1
    });

    console.log("\n=== wyszukiwanie semantyczne ===");
    results.documents[0]?.forEach(doc => console.log("-", doc));

    // Punkt 7: Filtrowanie przy użyciu metadanych
    query = "country";
    queryEmbedding = await getQueryEmbedding(query);

    results = await collection.query({
        queryEmbeddings: [queryEmbedding],
        nResults: 5,
        where: { continent: "Europe" } // Filtr: Szukaj tylko w Europie
    });

    console.log("\n=== państwa w Europie (filtrowanie po kontynencie) ===");
    results.documents[0]?.forEach(doc => console.log("-", doc));

    // --- Przykład filtrowania po konkretnej wartości (stolica) ---
    query = "capital cities";
    queryEmbedding = await getQueryEmbedding(query);

    results = await collection.query({
        queryEmbeddings: [queryEmbedding],
        nResults: 5,
        where: { capital: "Tokyo" }
    });

    console.log("\n=== filtrowanie po stolicy (Tokyo) ===");
    results.documents[0]?.forEach(doc => console.log("-", doc));

    // --- Przykład pokazujący nResults: 2 (Top 2 podobne państwa) ---
    query = "big country with many people in Asia";
    queryEmbedding = await getQueryEmbedding(query);

    results = await collection.query({
        queryEmbeddings: [queryEmbedding],
        nResults: 2 // Zadanie 5.4 wymaga znalezienia 2 podobnych państw
    });

    console.log("\n=== Top 2 podobne państwa ===");
    results.documents[0]?.forEach(doc => console.log("-", doc));
}

main().catch(console.error);